bundles/framework/layerlist/view/LayerViewTabs/LayerList/LayerCollapse/util.js

```diff
@@ -76,19 +116,130 @@ export const groupLayers = (layers, method, tools, allGroups = [], allDataProvid
                 groupForOrphans.addLayer(layer);
             }
         });
+    // recursively map groups and layers together
+    allGroups.map(parentGroup => {
+        group = mapGroupsAndLayers(parentGroup);
+        if (group != undefined) {
+            groupList.push(group);
+        }
+    });
 
-    let groupsWithoutLayers;
-    const lang = Oskari.getLang();
-    if (method === 'getInspireName') {
-        groupsWithoutLayers = allGroups.filter(t => groupList.filter(g => g.id === t.id).length === 0).map(t => {
+    let groupsWithoutLayers = [];
+    if (method !== 'getInspireName') {
+        groupsWithoutLayers = allDataProviders.filter(t => groupList.filter(g => g.id === t.id).length === 0).map(d => {
             const group = Oskari.clazz.create(
                 'Oskari.mapframework.bundle.layerselector2.model.LayerGroup',
-                t.id, method, t.name[lang]
+                d.id, method, d.name
             );
             group.setTools(tools);
             return group;
         });
-    } else {
+    }
+    groupsWithoutLayers = groupsWithoutLayers.sort((a, b) => Oskari.util.naturalSort(a.name, b.name));
+    return groupForOrphans ? [groupForOrphans, ...groupsWithoutLayers, ...groupList] : [...groupsWithoutLayers, ...groupList];
+};
+
+/**
+ * Function to construct layer groups based on information included in layers and given grouping method.
+ * Possible empty groups are included if allGroups and / or allDataProviders parameters are provided.
+ *
+ * @param {Oskari.mapframework.domain.AbstractLayer[]} layers layers to group
+ * @param {String} method layer method name to sort by
+ * @param {Oskari.mapframework.domain.Tool[]} tools tools to group
+ * @param {Oskari.mapframework.domain.MaplayerGroup[]} allGroups all user groups available in Oskari
+ * @param {Object[]} allDataProviders all dataproviders available in Oskari
+ */
+export const groupLayersAdmin = (layers, method, tools, allGroups = [], allDataProviders = [], noGroupTitle) => {
+    const groupList = [];
+    let group = null;
+    let groupForOrphans = null;
+
+    const determineGroupId = (layerGroups = [], layerAdmin) => {
+        let groupId;
+        if (method === 'getInspireName') {
+            if (layerGroups.length) {
+                groupId = layerGroups[0] ? layerGroups[0].id : undefined;
+            } else {
+                groupId = -1;
+            }
+        } else {
+            groupId = layerAdmin ? layerAdmin.organizationId : undefined;
+        }
+        // My map layers, my places, own analysis and 'orphan' groups don't have id so use negated random number
+        // as unique Id (with positive id group is interpret as editable and group tools are shown in layer list).
+        return typeof groupId === 'number' ? groupId : -Math.random();
+    };
+
+    const mapGroupsAndLayers = function (group) {
```

Comment:
Why is this a copy of mapGroupsAndLayers() from under groupLayers()? Why isn't the implementation shared?